st2common/st2common/util/crypto.py

```diff
@@ -13,40 +13,291 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+"""
+Module for handling symmetric encryption and decryption of short text values (mostly used for
+encrypted datastore values aka secrets).
+
+NOTE: In the past, this module used and relied on keyczar, but since keyczar doesn't support
+Python 3, we moved to cryptography library.
+
+symmetric_encrypt and symmetric_decrypt functions except values as returned by the AESKey.Encrypt()
+and AESKey.Decrypt() methods in keyczar. Those functions follow the same approach (AES in CBC mode
+with SHA1 HMAC signature) as keyczar methods, but they use and rely on primitives and methods from
+the cryptography library.
+
+This was done to make the keyczar -> cryptography migration fully backward compatible.
+
+Eventually, we should  move to Fernet (https://cryptography.io/en/latest/fernet/) recipe for
+symmetric encryption / decryption, because it offers more robustness and safer defaults (SHA256
+instead of SHA1, etc.).
+"""
+
 from __future__ import absolute_import
+
+import os
+import json
 import binascii
+import base64
+
+from hashlib import sha1
+
+import six
+
+from cryptography.hazmat.primitives.ciphers import Cipher
+from cryptography.hazmat.primitives.ciphers import algorithms
+from cryptography.hazmat.primitives.ciphers import modes
+from cryptography.hazmat.primitives import hashes
+from cryptography.hazmat.primitives import hmac
+from cryptography.hazmat.backends import default_backend
 
 __all__ = [
+    'KEYCZAR_HEADER_SIZE',
+    'KEYCZAR_AES_BLOCK_SIZE',
+    'KEYCZAR_HLEN',
+
     'read_crypto_key',
+
     'symmetric_encrypt',
-    'symmetric_decrypt'
+    'symmetric_decrypt',
+
+    'cryptography_symmetric_encrypt',
+    'cryptography_symmetric_decrypt',
+
+    # NOTE: Keyczar functions are here for testing reasons - they are only used by tests
```

Comment:
You don't need to export these explicitly if you tweak everything that imports them to dig them out of the module:

```python
from st2common.util.crypto import keyczar_symmetric_encrypt

keyczar_symmetric_encrypt(...)
```

can be rewritten as:

```python
import st2common.util.crypto as crypto

crypto.keyczar_symmetric_encrypt(...)
```

Reply:
Importing each function separately is just a personal preference. I prefer it over module imports where module names are generic as this one.

As far as exporting all the public functions and using ``__all__`` goes - that's just a "safety guard" and good practice I tend to follow. It makes it easier to see which public methods / classes a module exports and in case someone does "from foo import *", it will only include "public" methods. I know asterik imports are a bad idea and we should have a lint check for that, but still.

Reply:
Yep, I like to import each function separately too. Although I think you can still import individual functions even if they aren't "exported" with `__all__`, so you may not even need to export those functions at all. Whatever you think is best. 😃

Reply:
Yep, __all__ just affects asterisk import (``*``) import behavior :)